### DISCUSSION TITLE
Error: EMFILE: too many open files

### DISCUSSION THREAD

**Discussion by pkietz** [facebook/docusaurus#8812] [Q&A]
To start with: I have a very large Docusaurus project containing almost 10k .mdx files spread in 26 instances of `"@docusaurus/plugin-content-docs"`. It's been no problem to handle that amount of files in the development but now I have issues creating the production build.

The first issue I had to deal with was the `FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory` error. I've been able to solve it adding `NODE_OPTIONS=--max-old-space-size=16384` before starting `yarn build`.

But now I'm running into another issue right at the end of the production build proces:

```
$ NODE_OPTIONS=--max-old-space-size=16384 yarn build
yarn run v1.22.19
warning ..\package.json: No license field
$ docusaurus build
[INFO] [de] Creating an optimized production build...
i Compiling Client
i Compiling Server
√ Client: Compiled successfully in 50.20m
√ Server: Compiled successfully in 51.76m
[ERROR] Unable to build website for locale de.
[ERROR] Error: EMFILE: too many open files, open 'D:\path\path\build\path\filename.html'
[INFO] Docusaurus version: 2.3.1
Node version: v18.13.0
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Both, client and server, do compile successfuly, but due to the `EMFILE: too many open files` error 2 essential files are missing in the build folder: `sitemap.xml` and the `search-index.json` ("@easyops-cn/docusaurus-search-local"). I don't know exactly, which one of the 2 plugins is causing the error, the "Sitemap-plugin" or the "@easyops-cn/docusaurus-search-local"-plugin but it has be one of those - or maybe something's wrong with my local system (Windows 10, 64-bit, and I'm using GitBash as the prefered command line tool). 

Any ideas how to approach, to fix or bypass, this issue would be greatly appreciated.

Thank you very much.

**Comment by pkietz** [comment]
Quick update: I've temporarily disabled the search plugin in the production build and the build completed successfully. So the `EMFILE: too many open files` error seems to be caused by the "@easyops-cn/docusaurus-search-local"-plugin. 

If I understand correctly, the search plugin is build on top of algolia and algolia is dependend on the sitemap-plugin so there's probably no way to build the @easyops-cn search index with the sitemap deactivated. I'll have to use an other search solution in this case, I guess.

**Comment by slorber** [comment]
That's hard to help without a repro and I also don't use Windows.

Google says that multiple Windows users encounter this due to the Webpack Copy Plugin: https://github.com/webpack-contrib/copy-webpack-plugin

Maybe your project or one of your plugin uses this.

Refs: https://github.com/webpack-contrib/copy-webpack-plugin/issues/59

**Comment by pkietz** [comment]
Yeah, it seems to be a Windows specific issue and there's no easy way to deal with it by using `ulimit -n` in order to avoid any file descriptor limits during the build process. That's only working on Unix systems. I decided to move on with the issue unresolved by simply switching from the `easyops-cn` to the `docusaurus-lunr-search` plugin that is really nice as well and will do the job. The largest project I've been building so far using the `easyops-cn` search plugin had about 4k documents and the `easyops-cn` was able to handle it. It's good to know, that this plugin struggles when it comes to some crazy 10k docs, at least on Windows. The lunr plugin has been able to deal with the 10k docs without any issues.

Thank you very much!
